File: citywalk/legacy/www/server/application/scripts/models/content.py
#!/usr/local/bin/python
# -*- coding:utf-8 -*-
#
# models/content.py
#
# Content model
#

import pymongo
from bson import ObjectId
import datetime
from libcommon.enumlocale import EnumLocale
from libcommon.mongobase import MongoBase
from helpers.validator import Validator
from models.enums.media_type import MediaType
from models.enums.target_segment import TargetSegment


class Content(MongoBase):
    __collection__ = 'content'
    __structure__ = {
        '_id': ObjectId,
        'index': int,  # eg. 12
        'media_type': str,  # eg. MediaType.AUDIO_GUIDE.name

        # 'objective_id': ObjectId,  # to which objective is related

        'lat': float,
        'lon': float,

        'label': str,  # eg. Giza Pyramid
        'title': str,  # eg. The secret of Giza Pyramid
        'text': str,  # eg. The Great Pyramid of Giza is the oldest pyramids in the Giza pyramid complex

        'resource_key': str,  # eg. {organization_id}_{index}_{lang}.wav
        'image_key': str,  # eg. {organization_id}_{index}.jpg

        'language': str,  # eg. Language.ja.name

        'organization_id': str,  # The organization who registered this idea. ex) Organization._id

        'target': int,  # TargetSegment
        'radius': int,  # detection range
        'condition': dict,  # condition that this idea is reached. for instance, age restriction
        'featured': bool,  # featured flag
        'importance': int,  # publishing priority

        'created_member_id': ObjectId,  # _id of the member who created this idea  
        'latest_edit_member_id': ObjectId,  # _id of the member who last edited this idea  

        'deleted': bool,  # True when user chose deleting. after deleted, hidden but enable to recover.

        'created': datetime.datetime,
        'updated': datetime.datetime
    }
    __min_name_length__ = 1
    __max_name_length__ = 30
    __min_title_length__ = 3
    __max_title_length__ = 100
    __min_text_length__ = 20
    __max_text_length__ = 100
    __required_fields__ = ['_id', 'name', 'title', 'text', 'media_type', 'country']
    __default_values__ = {
        'deleted': False
    }
    __validators__ = {
        'name': Validator.validate_length('name', __min_name_length__, __max_name_length__),
        'title': Validator.validate_length('title', __min_title_length__, __max_title_length__),
        'text': Validator.validate_length('text', __min_text_length__, __max_text_length__),
    }
    __indexes__ = [
        [
            ("name", pymongo.HASHED),
        ]
    ]

    def response_json(self, excludes=[]):
        d = {}
        for key, val in self.items():
            if isinstance(val, ObjectId):
                val = str(val)
            d[key] = val
        return {x: d[x] for x in d if x not in excludes}

    @classmethod
    def fetch(cls, organization_id: ObjectId, with_deleted=False, json=True):
        """Fetch all ideas of the organization.

        args:
            - organization_id (ObjectId) : 
            - with_deleted (bool) : fetch deleted items if True
            - json (bool) : return by json if True

        returns:
            - ideas (list) : list of Idea objects
        """
        if with_deleted:
            ideas = cls.find({'organization_id': organization_id})
        else:
            ideas = cls.find({'organization_id': organization_id}, {"deleted": {"$ne": True}})
        if json:
            return [c.response_json() for c in ideas]
        else:
            return ideas

    @classmethod
    def validate_name_length(cls, name: str):
        if name < cls.__max_name_length__ and name > cls.__min_name_length__:
            return True
        else:
            return False

    @classmethod
    def validate_title_length(cls, title: str):
        if title < cls.__max_title_length__ and title > cls.__min_title_length__:
            return True
        else:
            return False

    @classmethod
    def validate_text_length(cls, text: str):
        if text < cls.__max_text_length__ and text > cls.__min_text_length__:
            return True
        else:
            return False

File: citywalk/legacy/www/server/application/views/src/ECMA/business/models/content.js
'use strict'
/**
 * @fileoverview business/models/content.js
 * Content data model class.
 * 
 * <code>
 * 
 * usage:
 *   var content = Content({
 *       _id: 'abcdefgh',
 *       ...
 *   });
 * 
 * functions:
 *   - validate()
 *       params:
 *       returns:
 *         {bool} - true if valid
 * 
 * </code>
 * 
 * @author kaz@thinkxinc.com (Kazuki Otsuka)
 */


/**
 * ContentType Enum.
 */
const ContentType = Object.freeze({
  audio_guide: 0,
  video: 1,
  others: 99,
});


/**
 * Content Data Model.
 * @param {string} _id -
 * @param {int} index -  
 * @param {string} type -  
 * @param {float} lat -  
 * @param {float} lon -  
 * @param {string} label -
 * @param {string} title -
 * @param {string} text -  
 * @param {string} resource_key -  
 * @param {string} image_key -  
 * @param {string} language -  
 * @param {string} organization_id -  
 * @param {dictionary} condition -  
 * @param {bool} featured -  
 * @param {int} importance -  
 * @param {string} created_member_id -  
 * @param {string} latest_edit_member_id -  
 * @param {bool} deleted -  
 */
class Content {
    _id = null
    index = null  
    type = null
    lat = null
    lon = null
    label = null
    title = null
    text = null
    resource_key = null
    image_key = null
    language = null
    organization_id = null
    target = null
    radius = null
    condition = null
    featured = null
    importance = null
    created_member_id = null
    latest_edit_member_id = null
    deleted = null
 
    constructor(
        _id, index, type,
        lat, lon,
        label, title, text,
        resource_key, image_key, language,
        organization_id,
        target, radius,
        condition, featured, importance, 
        created_member_id, latest_edit_member_id, deleted
        ) {
            this._id = _id;
            this.index = index;
            this.type = type;
            this.lat = lat;
            this.lon = lon;
            this.label = label;
            this.title = title;
            this.text = text;
            this.resource_key = resource_key;
            this.image_key = image_key;
            this.language = language;
            this.organization_id = organization_id;
            this.target = target
            this.radius = radius
            this.condition = condition;
            this.featured = featured;
            this.importance = importance;
            this.created_member_id = created_member_id;
            this.latest_edit_member_id = latest_edit_member_id;
            this.deleted = deleted;
   }

   static initFromJson(content) {
     // MEMO: somehow unable to set like _id=content._id in request callback.
     return new Content(
       (content._id != null) ? content._id : null,
       (content.index != null) ? content.index : null,
       (content.type != null) ? content.type : null,
       (content.lat != null) ? content.lat : null,
       (content.lon != null) ? content.lon : null,
       (content.label != null) ? content.label : null,
       (content.title != null) ? content.title : null,
       (content.text != null) ? content.text : null,
       (content.resource_key != null) ? content.resource_key : null,
       (content.image_key != null) ? content.image_key : null,
       (content.language != null) ? content.language : null,
       (content.organization_id != null) ? content.organization_id : null,
       (content.target != null) ? content.target : null,
       (content.radius != null) ? content.radius : null,
       (content.condition != null) ? content.condition : null,
       (content.featured != null) ? content.featured : null,
       (content.importance != null) ? content.importance : null,
       (content.created_member_id != null) ? content.created_member_id : null,
       (content.latest_edit_member_id != null) ? content.latest_edit_member_id : null,
       (content.deleted != null) ? content.deleted : null,
     )
   }

    /**
    * Minimum validation.
    * @param
    * @return {bool} true if valid.
    */
    validate() {

    }
}